### ISSUE TITLE
Using frozen_string_literal and unfreezing a string argument results in Ambiguous positive number operator warning

### ISSUE THREAD

**Issue by obfuscoder** [bbatsov/rubocop#4630]
## Expected behavior

With the recommended magic comment frozen_string_literal: true, which Rubocop checks for and autocorrects to, it is necessary to explitally mark the strings which shall not be frozen with +'...'.

I expect Rubocop not to complain about strings marked like this when the frozen_string_literal magic comment is defined.

## Actual behavior

Rubocop outputs a warning, when this string is defined as an argument to a method which mutates the string. In Rails 4, assert_select method mutates the pattern string when placeholder (?) are used:

```
W: Ambiguous positive number operator. Parenthesize the method arguments if it's surely a positive number operator, or add a whitespace to the right of the + if it should be a addition.
      assert_select +'input#item_price[name=?]', 'item[price]'
```

## Steps to reproduce the problem

1. Create file:

test.rb
```
# frozen_string_literal: true

puts +'frozen string'
```

2. Run rubocop test.rb

Output:

```
test.rb:3:6: W: Ambiguous positive number operator. Parenthesize the method arguments if it's surely a positive number operator, or add a whitespace to the right of the + if it should be a addition.
puts +'frozen string'
```

## Workaround

Using parenthes when calling the method helps:

```
puts(+'frozen string')
```

However I do not like to add parenthes there.

## RuboCop version

```
$ rubocop -V
0.49.1 (using Parser 2.4.0.0, running on ruby 2.4.1 x86_64-linux)
```

**Comment by rrosenblum** [comment]
@obfuscoder frozen string literal and ambiguous operator don't impact each other. `puts +'frozen string'` will complain about the ambiguous operator regardless of having frozen string literal enabled or not. 

The warning about the ambiguous operator is a redundant check to a Ruby warning. The issue is that it is difficult to tell whether `+` is an argument or an operator. As per your workaround, adding parentheses eliminates the ambiguity.

Notice the difference in invoking the code example with and without a space between the plus and quote.
```ruby
[3] pry(main)> puts +'a'
a
=> nil
[4] pry(main)> puts + 'a'

NoMethodError: undefined method `+' for nil:NilClass
from (pry):4:in `__pry__'
```


> Rubocop outputs a warning, when this string is defined as an argument to a method which mutates the string. In Rails 4, assert_select method mutates the pattern string when placeholder (?) are used:
```ruby
W: Ambiguous positive number operator. Parenthesize the method arguments if it's surely a positive number operator, or add a whitespace to the right of the + if it should be a addition.
      assert_select +'input#item_price[name=?]', 'item[price]'
```

RuboCop doesn't know that the method will be mutating the string argument. If frozen string literals is enabled, then Ruby will raise an exception when a string literal is mutated. In this example, the plus seems unnecessary regardless of frozen strings. If you are still having issues with this. Try using `'input#item_price[name=?]'.dup` or `String('input#item_price[name=?]')` to produce a mutable string.

Let me know if I can provide any other help for this issue.

**Comment by obfuscoder** [comment]
Hi @rrosenblum Thanks for your reply. As I mentioned, I'm not fond of using braces there to make RuboCop understand the intention. I try t avoid braces whenever possible. Your suggestions (#dup, String.new) are valid ones, but it would make the code even more bloated than braces. Unfortunatelly, I have to make those strings mutable as it is required by assert_select.

I suppose the actual issue is with assert_select and I will file create an issue for rails.

However I would appreciate an option in Rubocop which allowed me to disable the warning for +'...' when the frozen string literal is defined (or later when Ruby is handling all strings inmutable by default) as this is the recommended way of defining strings mutable.

**Comment by obfuscoder** [comment]
Created https://github.com/rails/rails-dom-testing/issues/69

**Comment by rrosenblum** [comment]
@obfuscoder thank you for creating a bug report with `rails-dom-testing`. I agree that `assert_select` needs to be modified to not mutate strings. Again, there is nothing special about the interaction with `Style/FrozenStringLiteralComment` and `Lint/AmbiguousOperator`. The error for the ambiguous operator would happen with or without the frozen string literal comment. You can disable the `Lint/AmbiguousOperator` cop to eliminate the warnings all together. I recommend leaving it enabled, because that cop mimics the functionality of Ruby's internal warnings.

```ruby
rrosenblum@C02NV0KSG3QN:~/work/rubocop (autocorrect_internal_affairs u=)$ cat test.rb 
puts +'frozen string'.freeze
puts +'frozen string'
puts(+'frozen string')
puts + 'frozen string'
rrosenblum@C02NV0KSG3QN:~/work/rubocop (autocorrect_internal_affairs u=)$ ruby -w test.rb 
test.rb:1: warning: ambiguous first argument; put parentheses or a space even after `+' operator
test.rb:2: warning: ambiguous first argument; put parentheses or a space even after `+' operator
test.rb:4: warning: possibly useless use of + in void context
frozen string
frozen string
frozen string

test.rb:4:in `<main>': undefined method `+' for nil:NilClass (NoMethodError)
```

**Comment by obfuscoder** [comment]
I haven't gotten any replies from the maintainers of rails_dom_testing yet. I hope to have this isse addressed some day. I'll close the issue on this side.
